### PR TITLE
allow users to set MaxDevices on KubeVirt CR

### LINKED ISSUE #6082: [Sig-scale] enable virt-operator to configure virt-handler to run more than 110 VMs per node
/kind enhancement

**Motivation**:
In the recent [SIG-scale meeting](https://docs.google.com/document/d/1d_b2o05FfBG37VwlC2Z1ZArnT9-_AEJoQTe7iKaQZ6I/edit#heading=h.50eclt5fctl9), we are discussing to run a scale test with more than 110 VMIs per node. It is trivial to increase the max number of pods per node, but the max number of VMIs is not straightforward. 

**What happened**:
Currently, `virt-handler` already have `max-devices` flag to increase the number of VMs per node
https://github.com/kubevirt/kubevirt/blob/0514a664151863d901d90694feb08c86df0076bb/cmd/virt-handler/virt-handler.go#L527

However, the `virt-handler` configuration is determined by the `virt-operator` that reverts all changes in the `virt-handler` Daemonset during the reconcile (of course, when we deploy it using the operator) .

**What you expected to happen**:
I expect to be able to configure `virt-operator` to create all `virt-handler` with a given max-devices number. Currently, all the `virt-handler` Daemonset options in `virt-operator` are hardcoded:
https://github.com/kubevirt/kubevirt/blob/0514a664151863d901d90694feb08c86df0076bb/pkg/virt-operator/resource/generate/components/daemonsets.go#L105

### PR DESCRIPTION
**Pull request by ashleyschuett**
**What this PR does / why we need it**:

Allows the user to easily modify the number of VMIs they would like to run per node

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6082

**Special notes for your reviewer**:

**Release note**:

```release-note
Add VirtualMachineInstancesPerNode to KubeVirt CR under Spec.Configuration
```

### CODE REVIEW

**Comment by iholder-redhat** [comment]
Thanks @ashleyschuett!
/lgtm

Note: I saw [this comment](https://github.com/kubevirt/kubevirt/blob/main/cmd/virt-handler/virt-handler.go#L527) in `virt-handler.go`.  Is this comment relevant? If we don't want to deprecate `max-devices` maybe we should delete it.

File: pkg/virt-operator/resource/apply/apps.go [hunk 1]
```diff
@@ -140,6 +145,20 @@ func (r *Reconciler) syncDaemonSet(daemonSet *appsv1.DaemonSet) error {
...
 }
 
+func setMaxDevices(kv *v1.KubeVirt, vh *appsv1.DaemonSet) {
+	maxDevices := getMaxDevices(kv)
+	if maxDevices == nil {
+		return
+	}
+
+	vh.Spec.Template.Spec.Containers[0].Command = append(vh.Spec.Template.Spec.Containers[0].Command, "--maxDevices")
+	vh.Spec.Template.Spec.Containers[0].Command = append(vh.Spec.Template.Spec.Containers[0].Command, fmt.Sprintf("%d", *maxDevices))
```
Comment (zcahana) [inline pkg/virt-operator/resource/apply/apps.go:155]: Can also fold to a single append call:
```go
vh.Spec.Template.Spec.Containers[0].Command = append(vh.Spec.Template.Spec.Containers[0].Command,
    "--maxDevices",
    fmt.Sprintf("%d", *maxDevices))
```

File: pkg/virt-operator/resource/apply/apps.go [hunk 2]
```diff
@@ -140,6 +145,20 @@ func (r *Reconciler) syncDaemonSet(daemonSet *appsv1.DaemonSet) error {
...
+}
+
+func getMaxDevices(kv *v1.KubeVirt) *int {
+	return kv.Spec.Configuration.MaxDevices
+}
```
Comment (zcahana) [inline pkg/virt-operator/resource/apply/apps.go:160]: I'm normally in favor of abstraction and extracting logic to separate methods, but I feel this particular case is somewhat redundant.
I think the following is perfectly fine:
```go
if kv.Spec.Configuration.MaxDevices == nil {
	return
}
command = append(command *kv.Spec.Configuration.MaxDevices)
```

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 3]
```diff
@@ -1879,6 +1879,7 @@ type KubeVirtConfiguration struct {
 	PermittedHostDevices        *PermittedHostDevices `json:"permittedHostDevices,omitempty"`
 	MinCPUModel                 string                `json:"minCPUModel,omitempty"`
 	ObsoleteCPUModels           map[string]bool       `json:"obsoleteCPUModels,omitempty"`
+	MaxDevices                  *int                  `json:"maxDevices,omitempty"`
 }
```
Comment (zcahana) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1883]: I realize "max devices" is the abstraction used internally by virt-handler, but it isn't the best terminology in the context
of a user-facing and cluster-wide terminology.

How about using "MaxVMsPerNode"? I think this better reflects the user's intent here, and also makes it clear that this is a per-node limit.

**Comment by rmohr** [comment]
I think with https://github.com/kubevirt/kubevirt/pull/6098 we have addressed the most urgent needs. However since this more flexible PR is already there, I would not mind having it.

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 3 above)
Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1883]: How about `virtualMachineInstancesPerNode`?

**Review by rmohr — COMMENTED** [review]
/approve

/hold

one additional suggestion to @zcahana regarding to a different name for the config setting.

**Comment by ashleyschuett** [comment]
Thanks for the reviews! I have updated it with the feedback :)

**Comment by zcahana** [comment]
> Thanks for the reviews! I have updated it with the feedback :)

Thanks @ashleyschuett!
/lgtm
/retest

I believe this can be /unholded, but I'll let you push the red button.

**Review by rmohr — COMMENTED** [review]
Looks good to me. Could you add a small e2e test which changes the values and wait to see the new values on the nodes?

**Pull request merged**